compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -611,6 +623,20 @@ impl<'brillig> Context<'brillig> {
     }
 }
 
+fn instruction_can_be_hoisted(
+    instruction: &Instruction,
+    dfg: &mut DataFlowGraph,
+    deduplicate_with_predicate: bool,
+) -> bool {
+    // These two can never be hoisted as they have a side-effect
+    // (though it's fine to de-duplicate them, just not fine to hoist them)
+    if matches!(instruction, Instruction::Constrain(..) | Instruction::RangeCheck { .. }) {
+        return false;
+    }
+
+    instruction.can_be_deduplicated(dfg, deduplicate_with_predicate)
```

Comment:
Another thing that messes with my head is what if a side effecting instruction is followed by a non-side effecting one that depends on its results: would we hoist the latter but not the former? Would it still make sense on its own?

Reply:
Good point, I'm not sure how that works.

Reply:
I guess the we consider the instructions together with their inputs, which are value IDs. If they were visible in both blocks, then the IDs must come from their dominators, otherwise they would be different, as the SSA makes sure all values created have a unique ID.